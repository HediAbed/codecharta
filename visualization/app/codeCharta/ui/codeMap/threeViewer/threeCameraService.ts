"use strict";

import * as THREE from "three";
import {
    SettingsServiceSubscriber,
    Settings,
    SettingsService
} from "../../../core/settings/settings.service";
import { PerspectiveCamera } from "three";

class ThreeCameraService implements SettingsServiceSubscriber {
    public static SELECTOR = "threeCameraService";

    public static VIEW_ANGLE = 45;
    public static NEAR = 100;
    public static FAR = 200000; //TODO optimize renderer for far objects

    public camera: PerspectiveCamera;

    constructor() {}

    onSettingsChanged(settings: Settings, event: Event) {
        this.setPosition(
            settings.camera.x,
            settings.camera.y,
            settings.camera.z
        );
    }

    init(
        settingsService: SettingsService,
        containerWidth: number,
        containerHeight: number,
        x: number,
        y: number,
        z: number
    ) {
        const aspect = containerWidth / containerHeight;
        this.camera = new THREE.PerspectiveCamera(
            ThreeCameraService.VIEW_ANGLE,
            aspect,
            ThreeCameraService.NEAR,
            ThreeCameraService.FAR
        );
        this.setPosition(x, y, z);
        settingsService.subscribe(this);
    }

    setPosition(x: number, y: number, z: number) {
        this.camera.position.set(x, y, z);
    }
}

export { ThreeCameraService };
