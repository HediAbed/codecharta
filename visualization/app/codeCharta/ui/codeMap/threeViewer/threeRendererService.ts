import { Camera, Scene, WebGLRenderer } from "three"
import { IRootScopeService } from "angular"
import { StoreService } from "../../../state/store.service"
import {
	IsWhiteBackgroundService,
	IsWhiteBackgroundSubscriber
} from "../../../state/store/appSettings/isWhiteBackground/isWhiteBackground.service"
import { CustomComposer } from "../rendering/postprocessor/customComposer"
import { RenderPass } from 'three/examples/jsm/postprocessing/RenderPass'
import { fxaaShaderStrings } from '../rendering/shaders/loaders/fxaaShaderString'
import { ShaderPass } from "three/examples/jsm/postprocessing/ShaderPass"
export class ThreeRendererService implements IsWhiteBackgroundSubscriber {
	
	static BACKGROUND_COLOR = {
		white: 0xffffff,
		normal: 0xeeeedd
	}

	static CLEAR_COLOR = ThreeRendererService.BACKGROUND_COLOR.normal

	static CLEAR_ALPHA = 1

	static RENDER_OPTIONS = {
		antialias: window.devicePixelRatio > 1.7 ? false : true,
									// 1.7 is just a guess number			
									// on deactivated pixel ratio>1.7 improves quality, 
									// performance hit is huge especially on fill rate limited gpus 
									// better to use fxaa and composing when device ratio is high
		preserveDrawingBuffer: true
	}

	composer: CustomComposer
	renderer: WebGLRenderer
	scene : Scene
	camera: Camera
	enableFXAA = false

	constructor(private storeService: StoreService, private $rootScope: IRootScopeService) {
		IsWhiteBackgroundService.subscribe(this.$rootScope, this)
	}

	init(containerWidth: number, containerHeight: number,scene : Scene, camera: Camera) {
		// eslint-disable-next-line no-console
		console.log("init")
		this.scene = scene
		this.camera = camera
		this.renderer = new WebGLRenderer(ThreeRendererService.RENDER_OPTIONS)
		this.renderer.setPixelRatio(window.devicePixelRatio)
		this.composer = new CustomComposer( this.renderer );
		this.renderer.setSize(containerWidth, containerHeight)
		this.onIsWhiteBackgroundChanged(this.storeService.getState().appSettings.isWhiteBackground)
		if (!ThreeRendererService.RENDER_OPTIONS.antialias) {
			this.enableFXAA = true
			this.initComposer()
		}
	}

	private initComposer = () => { 
		const pixelRatio = this.renderer.getPixelRatio()

		this.composer.setSize( window.innerWidth* pixelRatio, window.innerHeight* pixelRatio)
		const renderPass = new RenderPass( this.scene, this.camera )
		this.composer.addPass( renderPass )
		
		const effectFXAA = new ShaderPass( new fxaaShaderStrings() )
		effectFXAA.renderToScreen = false
        effectFXAA.uniforms[ 'resolution' ].value.x = 1 / ( window.innerWidth* pixelRatio ) 
		effectFXAA.uniforms[ 'resolution' ].value.y = 1 / ( window.innerHeight* pixelRatio )
        this.composer.addPass( effectFXAA )
	}

	onIsWhiteBackgroundChanged(isWhiteBackground: boolean) {
		if (isWhiteBackground) {
			ThreeRendererService.CLEAR_COLOR = ThreeRendererService.BACKGROUND_COLOR.white
		} else {
			ThreeRendererService.CLEAR_COLOR = ThreeRendererService.BACKGROUND_COLOR.normal
		}
		this.renderer.setClearColor(ThreeRendererService.CLEAR_COLOR, ThreeRendererService.CLEAR_ALPHA)
	}

	render() {
		const { scene, camera } = this
		if (this.enableFXAA) {
			this.composer.render()
		} else {
			this.renderer.render(scene, camera)
		}
	}
}
