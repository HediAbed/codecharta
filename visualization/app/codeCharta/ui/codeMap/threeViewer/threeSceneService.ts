import * as THREE from "three"
import { Scene } from "three"
import { Group } from "three"
import { CodeMapMesh } from "../rendering/codeMapMesh"
import { CodeMapBuilding } from "../rendering/codeMapBuilding"
import { Vector3 } from "three"
import { SettingsService } from "../../../state/settings.service"

/**
 * A service which manages the Three.js scene in an angular way.
 */
export class ThreeSceneService {
	public scene: Scene
	public labels: Group
	public edgeArrows: Group
	public mapGeometry: Group
	private lights: Group
	private mapMesh: CodeMapMesh

	constructor(private settingsService: SettingsService) {
		this.scene = new THREE.Scene()

		this.mapGeometry = new THREE.Group()
		this.lights = new THREE.Group()
		this.labels = new THREE.Group()
		this.edgeArrows = new THREE.Group()

		this.initLights()

		this.scene.add(this.mapGeometry)
		this.scene.add(this.edgeArrows)
		this.scene.add(this.labels)
		this.scene.add(this.lights)
	}

	public highlightBuilding(highlightedBuilding: CodeMapBuilding) {
		const isPresentationMode = this.settingsService.getSettings().appSettings.isPresentationMode
		const mapSize = this.settingsService.getSettings().treeMapSettings.mapSize

		for (let i = 0; i < this.mapMesh.getMeshDescription().buildings.length; i++) {
			const currentBuilding: CodeMapBuilding = this.mapMesh.getMeshDescription().buildings[i]
			const distance = highlightedBuilding.getCenterPoint(mapSize).distanceTo(currentBuilding.getCenterPoint(mapSize))

			if (currentBuilding.id !== highlightedBuilding.id) {
				if (isPresentationMode) {
					this.decreaseLightnessByDistance(currentBuilding, distance)
				} else {
					currentBuilding.decreaseLightness(20)
				}
			} else {
				currentBuilding.decreaseLightness(-10)
			}
			this.setVertexColor(currentBuilding.id, currentBuilding.getColorVector())
		}
		this.updateVertices()
	}

	public clearHighlight() {
		for (let i = 0; i < this.mapMesh.getMeshDescription().buildings.length; i++) {
			const currentBuilding: CodeMapBuilding = this.mapMesh.getMeshDescription().buildings[i]
			this.setVertexColor(currentBuilding.id, currentBuilding.getDefaultColorVector())
		}
		this.updateVertices()
	}

	private decreaseLightnessByDistance(building: CodeMapBuilding, distance: number) {
		if (distance > 800) {
			building.decreaseLightness(40)
		} else if (distance > 400) {
			building.decreaseLightness(30)
		} else if (distance > 250) {
			building.decreaseLightness(20)
		} else if (distance > 100) {
			building.decreaseLightness(15)
		} else if (distance > 50) {
			building.decreaseLightness(10)
		}
	}

	private setVertexColor(id: number, newColorVector: Vector3) {
		for (
			let j = id * CodeMapMesh.NUM_OF_VERTICES * CodeMapMesh.NUM_OF_COLOR_VECTOR_FIELDS;
			j <
			id * CodeMapMesh.NUM_OF_COLOR_VECTOR_FIELDS * CodeMapMesh.NUM_OF_VERTICES +
				CodeMapMesh.NUM_OF_COLOR_VECTOR_FIELDS * CodeMapMesh.NUM_OF_VERTICES;
			j += CodeMapMesh.NUM_OF_COLOR_VECTOR_FIELDS
		) {
			this.mapMesh.getThreeMesh().geometry["attributes"].color.array[j] = newColorVector.x
			this.mapMesh.getThreeMesh().geometry["attributes"].color.array[j + 1] = newColorVector.y
			this.mapMesh.getThreeMesh().geometry["attributes"].color.array[j + 2] = newColorVector.z
		}
	}

	private updateVertices() {
		this.mapMesh.getThreeMesh().geometry["attributes"].color.needsUpdate = true
	}

	public initLights() {
		const ambilight = new THREE.AmbientLight(0x707070) // soft white light
		const light1 = new THREE.DirectionalLight(0xe0e0e0, 1)
		light1.position.set(50, 10, 8).normalize()
		light1.castShadow = false
		light1.shadow.camera.right = 5
		light1.shadow.camera.left = -5
		light1.shadow.camera.top = 5
		light1.shadow.camera.bottom = -5
		light1.shadow.camera.near = 2
		light1.shadow.camera.far = 100

		const light2 = new THREE.DirectionalLight(0xe0e0e0, 1)
		light2.position.set(-50, 10, -8).normalize()
		light2.castShadow = false
		light2.shadow.camera.right = 5
		light2.shadow.camera.left = -5
		light2.shadow.camera.top = 5
		light2.shadow.camera.bottom = -5
		light2.shadow.camera.near = 2
		light2.shadow.camera.far = 100

		this.lights.add(ambilight)
		this.lights.add(light1)
		this.lights.add(light2)
	}

	public setMapMesh(mesh: CodeMapMesh, mapSize: number) {
		this.mapMesh = mesh

		while (this.mapGeometry.children.length > 0) {
			this.mapGeometry.remove(this.mapGeometry.children[0])
		}

		this.mapGeometry.position.x = -mapSize / 2.0
		this.mapGeometry.position.y = 0.0
		this.mapGeometry.position.z = -mapSize / 2.0

		this.mapGeometry.add(this.mapMesh.getThreeMesh())
	}

	public getMapMesh(): CodeMapMesh {
		return this.mapMesh
	}
}
