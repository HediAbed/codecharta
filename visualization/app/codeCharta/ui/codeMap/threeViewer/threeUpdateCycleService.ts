/**
 * This service allows other parts of the application to hook into the updateHovering cycle and get called on each cycle.
 */
export class ThreeUpdateCycleService {
	private updatables: Function[] = []

	public register(onUpdate: Function) {
		this.updatables.push(onUpdate)
	}

	/**
	 * Updates all registered callback functions
	 */
	public update() {
		this.updatables.forEach((u: Function) => {
			u()
		})
	}
}
