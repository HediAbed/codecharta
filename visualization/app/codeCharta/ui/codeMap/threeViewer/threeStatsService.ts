import Stats from 'three/examples/jsm/libs/stats.module';
import { WebGLInfo } from "three"
import { ThreeRendererService } from './threeRendererService';

export class ThreeStatsService {
	stats : Stats
    xPanel : Stats.Panel
	yPanel : Stats.Panel
	private maxXPanel = 0
    private maxYPanel = 0

    /* ngInject */
	constructor(
        private threeRendererService: ThreeRendererService
    ) {}

    init = (canvasElement: Element) => {
		const { stats = Stats() } = this
        
        this.xPanel = stats.addPanel( Stats.Panel( 'triangles', '#ff8', '#221' ) )
		this.yPanel = stats.addPanel( Stats.Panel( 'calls', '#f8f', '#212' ) )
		stats.showPanel( 3 )

		stats.domElement.style.position = 'absolute'
		stats.domElement.style.left = '0'
		stats.domElement.style.top = '0'
		canvasElement.appendChild( stats.dom )
    }

	updateStats = () => {
		const webGLInfo : WebGLInfo["render"] = this.threeRendererService.enableFXAA ? 
			this.threeRendererService.composer.getInfo() : 
			this.threeRendererService.renderer.info.render
		
		const triangles : number = webGLInfo.triangles
		this.maxXPanel = Math.max(this.maxXPanel,triangles)
		this.xPanel.update( triangles, this.maxXPanel*1.3 )

		const calls : number = webGLInfo.calls
		this.maxYPanel = Math.max(this.maxYPanel,calls)
		this.yPanel.update( calls, this.maxYPanel*1.3 )

		this.stats.update()
	}
}