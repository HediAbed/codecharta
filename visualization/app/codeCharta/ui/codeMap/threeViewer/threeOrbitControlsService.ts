import { ThreeCameraService } from "./threeCameraService"
import { IRootScopeService, IAngularEvent } from "angular"
import { Box3, CubeGeometry, Mesh, MeshNormalMaterial, OrbitControls, PerspectiveCamera, Vector3 } from "three"
import { ThreeSceneService } from "./threeSceneService"
import { SettingsService } from "../../../state/settingsService/settings.service"
import _ from "lodash"
import { FocusNodeSubscriber, UnfocusNodeSubscriber } from "../../../state/settingsService/settings.service.events"
import { LoadingGifService } from "../../loadingGif/loadingGif.service"

export interface CameraChangeSubscriber {
	onCameraChanged(camera: PerspectiveCamera)
}

/**
 * Service to manage the three orbit controls in an angular way.
 */
export class ThreeOrbitControlsService implements FocusNodeSubscriber, UnfocusNodeSubscriber {
	public static CAMERA_CHANGED_EVENT_NAME = "camera-changed"

	public controls: OrbitControls
	public defaultCameraPosition: Vector3 = new Vector3(0, 0, 0)
	public defaultZoom: number = 0

	/* ngInject */
	constructor(
		private threeCameraService: ThreeCameraService,
		private threeSceneService: ThreeSceneService,
		private $rootScope: IRootScopeService,
		private settingsService: SettingsService,
		private loadingGifService: LoadingGifService
	) {
		SettingsService.subscribeToFocusNode($rootScope, this)
		SettingsService.subscribeToUnfocusNode($rootScope, this)
	}

	public onFocusNode(focusPath: string) {
		this.autoFitTo()
	}

	public onUnfocusNode() {
		if (!this.loadingGifService.isLoadingNewFile() && !this.loadingGifService.isLoadingNewMap()) {
			this.resetCameraPerspective()
		}
	}

	public rotateCameraInVectorDirection(x: number, y: number, z: number) {
		const zoom = this.getZoom()
		this.lookAtDirectionFromTarget(x, y, z)
		this.applyOldZoom(zoom)
	}

	public cameraActionWhenNewMapIsLoaded() {
		if (this.settingsService.getSettings().appSettings.resetCameraIfNewFileIsLoaded) {
			this.autoFitTo()
			this.setDefaultZoom()
		} else {
			this.initializeDefaultZoomWithoutAutoFit()
		}
	}

	public autoFitTo() {
		const boundingSphere = this.getBoundingSphere()

		const len: number = this.autoFitToCalculation(boundingSphere)
		const cameraReference = this.threeCameraService.camera

		cameraReference.position.set(len, len, len)
		this.defaultCameraPosition = cameraReference.position.clone()
		this.controls.update()

		this.focusCameraViewToCenter(boundingSphere)
	}

	private resetCameraPerspective() {
		this.threeCameraService.camera.position
			.sub(this.controls.target)
			.setLength(this.defaultZoom)
			.add(this.controls.target)
	}

	private lookAtDirectionFromTarget(x: number, y: number, z: number) {
		this.threeCameraService.camera.position.set(this.controls.target.x, this.controls.target.y, this.controls.target.z)

		const alignmentCube = new Mesh(new CubeGeometry(20, 20, 20), new MeshNormalMaterial())

		this.threeSceneService.scene.add(alignmentCube)

		alignmentCube.position.set(this.controls.target.x, this.controls.target.y, this.controls.target.z)

		alignmentCube.translateX(x)
		alignmentCube.translateY(y)
		alignmentCube.translateZ(z)

		this.threeCameraService.camera.lookAt(alignmentCube.getWorldPosition())
		this.threeSceneService.scene.remove(alignmentCube)
	}

	private getZoom() {
		return this.threeCameraService.camera.position.distanceTo(this.controls.target)
	}

	private applyOldZoom(oldZoom: number) {
		this.threeCameraService.camera.translateZ(oldZoom)
	}

	private setDefaultZoom() {
		this.defaultZoom = this.getZoom()
	}

	private autoFitToCalculation(boundingSphere) {
		const cameraReference = this.threeCameraService.camera

		const scale = 1.4 // object size / display size
		const objectAngularSize = ((cameraReference.fov * Math.PI) / 180) * scale
		const distanceToCamera = boundingSphere.radius / Math.tan(objectAngularSize / 2)
		const len = Math.sqrt(Math.pow(distanceToCamera, 2) + Math.pow(distanceToCamera, 2))

		return len
	}

	private focusCameraViewToCenter(boundingSphere) {
		const t: Vector3 = boundingSphere.center.clone()
		t.setY(0)
		this.controls.target.set(t.x, t.y, t.z)

		this.threeCameraService.camera.lookAt(t)

		this.threeCameraService.camera.updateProjectionMatrix()
	}

	private initializeDefaultZoomWithoutAutoFit() {
		const autoFittedPerspective: Vector3 = new Vector3(0, 0, 0)
		const boundingSphere = this.getBoundingSphere()
		const len: number = this.autoFitToCalculation(boundingSphere)

		autoFittedPerspective.set(len, len, len)

		this.defaultCameraPosition = autoFittedPerspective.clone()

		const t: Vector3 = boundingSphere.center
		t.setY(0)

		const centerVector: Vector3 = new Vector3(t.x, t.y, t.z)

		this.defaultZoom = autoFittedPerspective.distanceTo(centerVector)
	}

	private getBoundingSphere() {
		return new Box3().setFromObject(this.threeSceneService.mapGeometry).getBoundingSphere()
	}

	public init(domElement) {
		const OrbitControls = require("three-orbit-controls")(require("three"))
		this.controls = new OrbitControls(this.threeCameraService.camera, domElement)
		this.controls.addEventListener("change", () => {
			this.onInput(this.threeCameraService.camera)
		})
	}

	public onInput(camera: PerspectiveCamera) {
		this.$rootScope.$broadcast(ThreeOrbitControlsService.CAMERA_CHANGED_EVENT_NAME, camera)
	}

	public static subscribe($rootScope: IRootScopeService, subscriber: CameraChangeSubscriber) {
		$rootScope.$on(ThreeOrbitControlsService.CAMERA_CHANGED_EVENT_NAME, (event: IAngularEvent, camera: PerspectiveCamera) => {
			subscriber.onCameraChanged(camera)
		})
	}
}
