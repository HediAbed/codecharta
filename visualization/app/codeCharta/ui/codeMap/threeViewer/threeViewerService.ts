"use strict"

import * as Stats from "stats-js"	// NOTE npm i stats-js
import { ThreeSceneService } from "./threeSceneService"
import { ThreeCameraService } from "./threeCameraService"
import { ThreeOrbitControlsService } from "./threeOrbitControlsService"
import { ThreeRendererService } from "./threeRendererService"
import { ThreeUpdateCycleService } from "./threeUpdateCycleService"
import { RenderPass } from 'three/examples/jsm/postprocessing/RenderPass'
import { fxaaShaderStrings } from '../rendering/fxaaShaderString'
import { ShaderPass } from "three/examples/jsm/postprocessing/ShaderPass"
import { WebGLInfo } from "three"

export class ThreeViewerService {
	private stats : Stats
	private xPanel
	private yPanel
	private maxXPanel = 0
	private maxYPanel = 0
	private enableFXAA = false // TODO can be selective

	/* ngInject */
	constructor(
		private threeSceneService: ThreeSceneService,
		private threeCameraService: ThreeCameraService,
		private threeOrbitControlsService: ThreeOrbitControlsService,
		private threeRendererService: ThreeRendererService,
		private threeUpdateCycleService: ThreeUpdateCycleService
	) {
		this.stats = new Stats()	// TODO needs to be injected only for test purpose 
		
		/* 	NOTE FXAA instead of aliasing shader set to medium quality,
			could try out different filtering techniques */
		if (!ThreeRendererService.RENDER_OPTIONS.antialias) {
			this.enableFXAA = true
		}							
	}

	init(canvasElement: Element) {
		this.initStats(canvasElement)
		this.threeCameraService.init(window.innerWidth, window.innerHeight)
		this.threeCameraService.camera.lookAt(this.threeSceneService.scene.position)
		this.threeSceneService.scene.add(this.threeCameraService.camera)
		this.threeRendererService.init(window.innerWidth, window.innerHeight)
		this.threeOrbitControlsService.init(this.threeRendererService.renderer.domElement)
		this.threeRendererService.renderer.setPixelRatio(window.devicePixelRatio)  // TODO needs ui progressbar (from 1 to window.devicePixelRatio)
		
		if (this.enableFXAA) {
			this.initComposer()
		}

		canvasElement.appendChild(this.threeRendererService.renderer.domElement)

		window.addEventListener("resize", () => this.onWindowResize())
		window.addEventListener("focusin", event => this.onFocusIn(event))
		window.addEventListener("focusout", event => this.onFocusOut(event))
	}

	initStats = (canvasElement: Element) => {
		// TODO refactoring is needed
		const { stats } = this
		this.xPanel = stats.addPanel( new Stats.Panel( 'triangles', '#ff8', '#221' ) )
		this.yPanel = stats.addPanel( new Stats.Panel( 'calls', '#f8f', '#212' ) )
		stats.showPanel( 3 )

		stats.domElement.style.position = 'absolute'
		stats.domElement.style.left = '0'
		stats.domElement.style.top = '0'
		canvasElement.appendChild( stats.dom )
	}

	private updateStats = () => {
		const webGLInfo : WebGLInfo["render"]= this.enableFXAA ? 
			this.threeRendererService.composer.getInfo() : 
			this.threeRendererService.renderer.info.render
		
		const triangles : number = webGLInfo.triangles
		this.maxXPanel = Math.max(this.maxXPanel,triangles)
		this.xPanel.update( triangles, this.maxXPanel*1.3 )

		const calls : number = webGLInfo.calls
		this.maxYPanel = Math.max(this.maxYPanel,calls)
		this.yPanel.update( calls, this.maxYPanel*1.3 )

		this.stats.update()
	}

	// TODO render pass needs to be injected
	private initComposer = () => { 
		const pixelRatio = this.threeRendererService.renderer.getPixelRatio()

		this.threeRendererService.composer.setSize( window.innerWidth* pixelRatio, window.innerHeight* pixelRatio)
		const renderPass = new RenderPass( this.threeSceneService.scene, this.threeCameraService.camera )
		this.threeRendererService.composer.addPass( renderPass )
		
		const effectFXAA = new ShaderPass( new fxaaShaderStrings() )
		effectFXAA.renderToScreen = false
        effectFXAA.uniforms[ 'resolution' ].value.x = 1 / ( window.innerWidth* pixelRatio ) 
		effectFXAA.uniforms[ 'resolution' ].value.y = 1 / ( window.innerHeight* pixelRatio )
        this.threeRendererService.composer.addPass( effectFXAA )
	}

	onWindowResize() {
		this.threeSceneService.scene.updateMatrixWorld(false)
		this.threeRendererService.renderer.setSize(window.innerWidth, window.innerHeight)
		this.threeCameraService.camera.aspect = window.innerWidth / window.innerHeight
		this.threeCameraService.camera.updateProjectionMatrix()
	}

	onFocusIn(event) {
		if (event.target.nodeName === "INPUT") {
			this.threeOrbitControlsService.controls.enableKeys = false
		}
	}

	onFocusOut(event) {
		if (event.target.nodeName === "INPUT") {
			this.threeOrbitControlsService.controls.enableKeys = true
		}
	}

	private render() {
		if (this.enableFXAA) {
			this.threeRendererService.composer.render()
		} else {
			this.threeRendererService.renderer.render(this.threeSceneService.scene, this.threeCameraService.camera)
		}
	}

	dispose() {
		this.threeRendererService?.composer?.dispose()
	}

	animate() {
		requestAnimationFrame(() => this.animate())
		this.render()
		this.threeOrbitControlsService.controls.update()
		this.threeUpdateCycleService.update()
		this.updateStats()
	}
}
