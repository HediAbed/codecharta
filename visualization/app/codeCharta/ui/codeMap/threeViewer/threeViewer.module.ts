import {ThreeViewerService} from "./threeViewerService";
import {ThreeSceneService} from "./threeSceneService";
import {ThreeCameraService} from "./threeCameraService";
import {ThreeOrbitControlsService} from "./threeOrbitControlsService";
import {ThreeRendererService} from "./threeRendererService";
import {ThreeUpdateCycleService} from "./threeUpdateCycleService";
import "../../../state/state.module";
import _ from "lodash"

import angular from "angular";

angular.module("app.codeCharta.ui.codeMap.threeViewer",
    ["app.codeCharta.state"]
).service(
    _.camelCase(ThreeViewerService.name),
    ThreeViewerService
).service(
    _.camelCase(ThreeUpdateCycleService.name),
    ThreeUpdateCycleService
).service(
    _.camelCase(ThreeSceneService.name),
    ThreeSceneService
).service(
    _.camelCase(ThreeRendererService.name),
    ThreeRendererService
).service(
    _.camelCase(ThreeOrbitControlsService.name),
    ThreeOrbitControlsService
).service(
    _.camelCase(ThreeCameraService.name),
    ThreeCameraService
);


