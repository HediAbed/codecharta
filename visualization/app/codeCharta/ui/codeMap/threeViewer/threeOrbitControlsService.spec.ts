import "./threeViewer.module"
import { getService, instantiateModule } from "../../../../../mocks/ng.mockhelper"
import { ThreeOrbitControlsService } from "./threeOrbitControlsService"
import { ThreeCameraService } from "./threeCameraService"
import { ThreeSceneService } from "./threeSceneService"
import { IRootScopeService } from "angular"
import * as THREE from "three"
import { OrbitControls, PerspectiveCamera, Vector3 } from "three"
import { SettingsService } from "../../../state/settingsService/settings.service"
import { LoadingGifService } from "../../loadingGif/loadingGif.service"

describe("ThreeOrbitControlsService", () => {
	let threeCameraService: ThreeCameraService
	let threeSceneService: ThreeSceneService
	let $rootScope: IRootScopeService
	let threeOrbitControlsService: ThreeOrbitControlsService
	let settingsService: SettingsService
	let loadingGifService: LoadingGifService

	let vector: Vector3

	afterEach(() => {
		jest.resetAllMocks()
	})

	beforeEach(() => {
		restartSystem()
		rebuildService()
		withMockedThreeCameraService()
		withMockedThreeSceneService()
	})

	function restartSystem() {
		instantiateModule("app.codeCharta.ui.codeMap.threeViewer")

		threeCameraService = getService<ThreeCameraService>("threeCameraService")
		threeSceneService = getService<ThreeSceneService>("threeSceneService")
		$rootScope = getService<IRootScopeService>("$rootScope")
		settingsService = getService<SettingsService>("settingsService")
		loadingGifService = getService<LoadingGifService>("loadingGifService")

		vector = new Vector3(4.4577067775672665, 4.4577067775672665, 4.4577067775672665)
	}

	function withMockedThreeCameraService() {
		const camera = new PerspectiveCamera(100, 0, 0, 0)
		camera.position.set(vector.x, vector.y, vector.z)
		threeCameraService = threeOrbitControlsService["threeCameraService"] = jest.fn<ThreeCameraService>(() => {
			return {
				camera
			}
		})()
	}

	function withMockedThreeSceneService() {
		threeSceneService = threeOrbitControlsService["threeSceneService"] = jest.fn<ThreeSceneService>(() => {
			return {
				scene: {
					add: jest.fn(),
					remove: jest.fn()
				},
				mapGeometry: new THREE.Mesh(new THREE.BoxGeometry(10, 10, 10))
			}
		})()
	}

	function rebuildService() {
		threeOrbitControlsService = new ThreeOrbitControlsService(
			threeCameraService,
			threeSceneService,
			$rootScope,
			settingsService,
			loadingGifService
		)
	}

	describe("constructor", () => {
		it("should subscribe to FocusedNodePath-Event", () => {
			SettingsService.subscribeToFocusNode = jest.fn()

			rebuildService()

			expect(SettingsService.subscribeToFocusNode).toHaveBeenCalledWith($rootScope, threeOrbitControlsService)
		})
	})

	it("rotateCameraInVectorDirection ", () => {
		threeOrbitControlsService.controls = {
			target: new THREE.Vector3(0, 0, 0)
		} as OrbitControls
		const vector = { x: 0, y: 1, z: 0 }

		threeOrbitControlsService.rotateCameraInVectorDirection(vector.x, vector.y, vector.z)

		expect(threeSceneService.scene.add).toBeCalled()
		expect(threeSceneService.scene.remove).toBeCalled()

		expect(threeCameraService.camera.position).toMatchSnapshot()
	})

	describe("onFocusedNodePathChanged", () => {
		beforeEach(() => {
			threeOrbitControlsService.controls = {
				target: new THREE.Vector3(1, 1, 1)
			} as OrbitControls
			threeOrbitControlsService.controls.update = jest.fn()
		})

		it("should set the camera perspective, to the origin value", () => {
			threeOrbitControlsService.defaultCameraPosition.set(12, 13, 14)

			threeOrbitControlsService.onFocusNode("")

			expect(threeCameraService.camera.position).toMatchSnapshot()
		})

		it("should set the camera perspective, to Vector with 0 if no default Value is saved", () => {
			threeOrbitControlsService.onFocusNode("")

			expect(threeCameraService.camera.position).toMatchSnapshot()
		})

		it("autoFitTo ", () => {
			threeOrbitControlsService.onFocusNode("something")

			expect(threeOrbitControlsService.controls.update).toBeCalled()

			expect(threeCameraService.camera.position).toMatchSnapshot()
		})
	})
})
