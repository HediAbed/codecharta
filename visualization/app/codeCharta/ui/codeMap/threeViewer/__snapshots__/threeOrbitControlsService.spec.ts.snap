// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThreeOrbitControlsService rotateCameraInVectorDirection  1`] = `
Vector3 {
  "x": 0,
  "y": -7.720974585385768,
  "z": 0.0007720974585381485,
}
`;
